File: python/comms/agents.py
"""
Agent classes for the agent-to-agent communication loop.
"""
from __future__ import annotations

import sys
import os

_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
if _DIR not in sys.path:
    sys.path.insert(0, _DIR)

from agent.llm import get_response_from_llm


# ── helpers ──────────────────────────────────────────────────────────────────

def _word_count(text: str) -> int:
    return len(text.split())


def _make_history(system_prompt: str) -> list:
    """Start a fresh conversation history with a system message."""
    return [{"role": "system", "content": system_prompt}]


# ── CommunicatingAgent ────────────────────────────────────────────────────────

class CommunicatingAgent:
    """
    One side of the two-agent communication pair.

    Each round the agent is given a fresh history seeded with its system prompt
    (which includes the task context and accumulated overseer tips).
    """

    def __init__(self, name: str, model: str):
        self.name = name
        self.model = model
        self.history: list = []
        self.tips: list[str] = []          # tips from overseer across rounds
        self.word_counts: list[int] = []   # words per message this round

    # ── round lifecycle ───────────────────────────────────────────────────────

    def start_round(self, system_prompt: str) -> None:
        """Reset conversation history for a new round."""
        full_prompt = system_prompt
        if self.tips:
            tips_block = "\n".join(f"  • {t}" for t in self.tips)
            full_prompt += (
                f"\n\nOVERSEER TIPS FROM PREVIOUS ROUNDS (apply these now):\n{tips_block}"
            )
        self.history = _make_history(full_prompt)
        self.word_counts = []

    def receive(self, message: str) -> str:
        """
        Receive a message from the other agent and generate a reply.
        Returns the reply text.
        """
        reply, self.history, _ = get_response_from_llm(
            msg=message,
            model=self.model,
            msg_history=self.history,
        )
        self.word_counts.append(_word_count(reply))
        return reply

    def add_tip(self, tip: str) -> None:
        """Accumulate overseer tip for future rounds."""
        self.tips.append(tip)

    @property
    def avg_words(self) -> float:
        if not self.word_counts:
            return 0.0
        return sum(self.word_counts) / len(self.word_counts)


# ── OverseerAgent ─────────────────────────────────────────────────────────────

OVERSEER_SYSTEM = """\
You are the Overseer of a two-agent communication experiment.
Your job after each round:

1. READ the full exchange between Agent A and Agent B.
2. SCORE the round's communication efficiency on a scale 1–10 where:
   • 1 = extremely verbose, full of filler, redundancy, hedging
   • 5 = adequate but room to improve
   • 10 = perfectly compressed, every word carries signal, zero waste
3. Give SPECIFIC TIPS to each agent (2–3 bullet points each):
   • Point to exact phrases that were wasteful or unclear
   • Suggest concrete alternatives or structural improvements
   • Encourage developing shorthand, abbreviations, or schemas if appropriate
4. If the task has a verification question, VERIFY whether the information
   was transmitted correctly and completely.

Respond in this exact format:
SCORE: <1–10>
VERDICT: <one sentence on whether information was transmitted correctly>
TIPS FOR AGENT A:
• <tip 1>
• <tip 2>
• <tip 3 if needed>
TIPS FOR AGENT B:
• <tip 1>
• <tip 2>
• <tip 3 if needed>
"""


class OverseerAgent:
    """
    Third-party agent that watches the A↔B exchange and coaches both sides.
    """

    def __init__(self, model: str):
        self.model = model
        self.scores: list[int] = []
        self.history: list = []

    def evaluate(
        self,
        exchange_log: list[dict],  # [{"agent": "A"|"B", "text": "..."}]
        task_context: str = "",
        verification_questions: list[str] | None = None,
        agent_b_last_reply: str = "",
    ) -> dict:
        """
        Evaluate one round of exchange and return parsed feedback.

        Returns:
            {
                "score": int,
                "verdict": str,
                "tips_a": list[str],
                "tips_b": list[str],
                "raw": str,
            }
        """
        # Build prompt
        exchange_text = "\n\n".join(
            f"[Agent {e['agent']}]: {e['text']}" for e in exchange_log
        )
        prompt_parts = []
        if task_context:
            prompt_parts.append(f"TASK CONTEXT:\n{task_context}")
        prompt_parts.append(f"EXCHANGE:\n{exchange_text}")
        if verification_questions:
            qs = "\n".join(f"  {i+1}. {q}" for i, q in enumerate(verification_questions))
            prompt_parts.append(
                f"VERIFICATION — check Agent B's replies against these questions:\n{qs}\n"
                f"Agent B's last reply: {agent_b_last_reply}"
            )
        prompt = "\n\n".join(prompt_parts)

        raw, self.history, _ = get_response_from_llm(
            msg=prompt,
            model=self.model,
            msg_history=_make_history(OVERSEER_SYSTEM),
        )
        return self._parse(raw)

    def _parse(self, raw: str) -> dict:
        result = {"score": 5, "verdict": "", "tips_a": [], "tips_b": [], "raw": raw}
        section = None
        for line in raw.splitlines():
            stripped = line.strip()
            if stripped.startswith("SCORE:"):
                try:
                    result["score"] = int("".join(c for c in stripped.split(":", 1)[1] if c.isdigit() or c == ".").split(".")[0])
                except (ValueError, IndexError):
                    pass
                self.scores.append(result["score"])
            elif stripped.startswith("VERDICT:"):
                result["verdict"] = stripped.split(":", 1)[1].strip()
            elif "TIPS FOR AGENT A" in stripped.upper():
                section = "a"
            elif "TIPS FOR AGENT B" in stripped.upper():
                section = "b"
            elif stripped.startswith("•") or stripped.startswith("-"):
                tip = stripped.lstrip("•- ").strip()
                if tip:
                    if section == "a":
                        result["tips_a"].append(tip)
                    elif section == "b":
                        result["tips_b"].append(tip)
        return result

File: rust/src/comms/agents.rs
use anyhow::Result;
use crate::llm::{LlmClient, Message};

// ── helpers ───────────────────────────────────────────────────────────────────

fn word_count(s: &str) -> usize {
    s.split_whitespace().count()
}

fn make_history(system_prompt: &str) -> Vec<Message> {
    vec![Message { role: "system".into(), content: system_prompt.into() }]
}

// ── ExchangeEntry ─────────────────────────────────────────────────────────────

#[derive(Clone)]
pub struct ExchangeEntry {
    pub agent: char,   // 'A' or 'B'
    pub text: String,
}

// ── CommunicatingAgent ────────────────────────────────────────────────────────

pub struct CommunicatingAgent {
    pub name: char,
    client: LlmClient,
    history: Vec<Message>,
    pub tips: Vec<String>,
    pub word_counts: Vec<usize>,
}

impl CommunicatingAgent {
    pub fn new(name: char, model: &str) -> Self {
        Self {
            name,
            client: LlmClient::new(model),
            history: vec![],
            tips: vec![],
            word_counts: vec![],
        }
    }

    /// Reset conversation history for a new round, prepending accumulated tips.
    pub fn start_round(&mut self, system_prompt: &str) {
        let mut full = system_prompt.to_string();
        if !self.tips.is_empty() {
            full.push_str("\n\nOVERSEER TIPS FROM PREVIOUS ROUNDS (apply these now):\n");
            for tip in &self.tips {
                full.push_str(&format!("  • {tip}\n"));
            }
        }
        self.history = make_history(&full);
        self.word_counts.clear();
    }

    /// Send `message` to this agent and return its reply.
    pub fn receive(&mut self, message: &str) -> Result<String> {
        self.history.push(Message { role: "user".into(), content: message.into() });
        let reply = self.client.chat(&self.history)?;
        self.history.push(Message { role: "assistant".into(), content: reply.clone() });
        self.word_counts.push(word_count(&reply));
        Ok(reply)
    }

    pub fn add_tip(&mut self, tip: String) {
        self.tips.push(tip);
    }

    pub fn avg_words(&self) -> f64 {
        if self.word_counts.is_empty() {
            return 0.0;
        }
        self.word_counts.iter().sum::<usize>() as f64 / self.word_counts.len() as f64
    }
}

// ── OverseerAgent ─────────────────────────────────────────────────────────────

const OVERSEER_SYSTEM: &str = "\
You are the Overseer of a two-agent communication experiment.
Your job after each round:

1. READ the full exchange between Agent A and Agent B.
2. SCORE the round's communication efficiency on a scale 1–10 where:
   • 1 = extremely verbose, full of filler, redundancy, hedging
   • 5 = adequate but room to improve
   • 10 = perfectly compressed, every word carries signal, zero waste
3. Give SPECIFIC TIPS to each agent (2–3 bullet points each):
   • Point to exact phrases that were wasteful or unclear
   • Suggest concrete alternatives or structural improvements
   • Encourage developing shorthand, abbreviations, or schemas if appropriate
4. If the task has a verification question, VERIFY whether the information
   was transmitted correctly and completely.

Respond in this exact format:
SCORE: <1–10>
VERDICT: <one sentence on whether information was transmitted correctly>
TIPS FOR AGENT A:
• <tip 1>
• <tip 2>
• <tip 3 if needed>
TIPS FOR AGENT B:
• <tip 1>
• <tip 2>
• <tip 3 if needed>";

pub struct OverseerFeedback {
    pub score: u8,
    pub verdict: String,
    pub tips_a: Vec<String>,
    pub tips_b: Vec<String>,
    pub raw: String,
}

pub struct OverseerAgent {
    client: LlmClient,
    pub scores: Vec<u8>,
}

impl OverseerAgent {
    pub fn new(model: &str) -> Self {
        Self { client: LlmClient::new(model), scores: vec![] }
    }

    pub fn evaluate(
        &mut self,
        exchange_log: &[ExchangeEntry],
        task_context: &str,
        verification_questions: &[&str],
        agent_b_last_reply: &str,
    ) -> Result<OverseerFeedback> {
        let exchange_text: String = exchange_log
            .iter()
            .map(|e| format!("[Agent {}]: {}", e.agent, e.text))
            .collect::<Vec<_>>()
            .join("\n\n");

        let mut parts = vec![];
        if !task_context.is_empty() {
            parts.push(format!("TASK CONTEXT:\n{task_context}"));
        }
        parts.push(format!("EXCHANGE:\n{exchange_text}"));
        if !verification_questions.is_empty() {
            let qs = verification_questions
                .iter()
                .enumerate()
                .map(|(i, q)| format!("  {}. {q}", i + 1))
                .collect::<Vec<_>>()
                .join("\n");
            parts.push(format!(
                "VERIFICATION — check Agent B's replies against these questions:\n{qs}\n\
                 Agent B's last reply: {agent_b_last_reply}"
            ));
        }

        let prompt = parts.join("\n\n");
        let messages = [
            Message { role: "system".into(), content: OVERSEER_SYSTEM.into() },
            Message { role: "user".into(), content: prompt },
        ];

        let raw = self.client.chat(&messages)?;
        let fb = self.parse(&raw);
        self.scores.push(fb.score);
        Ok(fb)
    }

    fn parse(&self, raw: &str) -> OverseerFeedback {
        let mut score: u8 = 5;
        let mut verdict = String::new();
        let mut tips_a: Vec<String> = vec![];
        let mut tips_b: Vec<String> = vec![];
        let mut section = ' ';

        for line in raw.lines() {
            let s = line.trim();
            if let Some(rest) = s.strip_prefix("SCORE:") {
                let digits: String = rest.chars().filter(|c| c.is_ascii_digit()).collect();
                score = digits.parse::<u8>().unwrap_or(5).min(10).max(1);
            } else if let Some(rest) = s.strip_prefix("VERDICT:") {
                verdict = rest.trim().to_string();
            } else if s.to_uppercase().contains("TIPS FOR AGENT A") {
                section = 'A';
            } else if s.to_uppercase().contains("TIPS FOR AGENT B") {
                section = 'B';
            } else if s.starts_with('•') || s.starts_with('-') {
                let tip = s.trim_start_matches(['•', '-', ' ']).trim().to_string();
                if !tip.is_empty() {
                    match section {
                        'A' => tips_a.push(tip),
                        'B' => tips_b.push(tip),
                        _ => {}
                    }
                }
            }
        }

        OverseerFeedback { score, verdict, tips_a, tips_b, raw: raw.to_string() }
    }
}
